store/postgres/src/mapping_sql.rs

```diff
@@ -0,0 +1,1107 @@
+///! This module contains the gory details of using Diesel to query
+///! a database schema that is not known at compile time. The code in this
+///! module is mostly concerned with constructing SQL queries and some
+///! helpers for serializing and deserializing entities.
+///!
+///! Code in this module works very hard to minimize the number of allocations
+///! that it performs
+use diesel::pg::{Pg, PgConnection};
+use diesel::query_builder::{AstPass, QueryFragment, QueryId};
+use diesel::query_dsl::{LoadQuery, RunQueryDsl};
+use diesel::result::QueryResult;
+use diesel::sql_types::{Array, Binary, Bool, Integer, Jsonb, Numeric, Range, Text};
+use diesel::Connection;
+use failure::Fail;
+use std::convert::TryFrom;
+use std::str::FromStr;
+
+use graph::data::store::scalar;
+use graph::prelude::{
+    format_err, serde_json, Attribute, Entity, EntityFilter, EntityKey, StoreError, Value,
+    ValueType,
+};
+
+use crate::block_range::{BlockNumber, BlockRange, BlockRangeContainsClause};
+use crate::filter::UnsupportedFilter;
+use crate::mapping::{
+    Column, ColumnType, Mapping, SqlName, Table, BLOCK_RANGE, PRIMARY_KEY_COLUMN,
+};
+use crate::sql_value::SqlValue;
+
+/// Helper struct for retrieving entities from the database. With diesel, we
+/// can only run queries that return columns whose number and type are known
+/// at compile time. Because of that, we retrieve the actual data for an
+/// entity as Jsonb by converting the row containing the entity using the
+/// `to_jsonb` function.
+#[derive(QueryableByName)]
+pub struct EntityData {
+    #[sql_type = "Text"]
+    entity: String,
+    #[sql_type = "Jsonb"]
+    data: serde_json::Value,
```

Comment:
That would be nicer, but doesn't work because
```
the trait `diesel::deserialize::FromSql<diesel::sql_types::Jsonb, Pg>`
is not implemented for `serde_json::Map<String, serde_json::Value>`
```

We know that we will always get an object back (since we use `to_json` in SQL), but can't express that well in the type system

Reply:
I see, thanks for checking.